(closes #1499) Method to resolve properties of symbols imported from wildcard imports

sergisiso opened the pull request.

sergisiso commented:
@arporter @rupertford This is ready for review. Codecov marks a missing coverage line that has me quite confused, if I replace the line with pdb.set_trace() one of the tests stops on it, and if I replace it with a "raise" and send it to the CI the test does fail as expected. So I am not sure why the CodeCov is not happy with it. Otherwise the PR is ready for a first look.

arporter commented on src/psyclone/tests/psyir/symbols/data_type_symbol_test.py:
```diff
@@ -72,3 +72,22 @@ def test_datatypesymbol_copy():
+
+    # new_sym name should be unchanged, but its datatype should be updated
+    assert new_sym.name == "new_name"
+    assert new_sym.datatype == symbol.datatype
```
As commented in the implementation, I'm concerned that the `datatype` of `new_sym` is now a reference to the `datatype` of `symbol`?

sergisiso replied:
Yes, but this is similar to what we do with `== INTEGER_TYPE`, so if it is not what we want/need this needs a redesign expanding outside the scope of this PR.

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+            Symbol("not_used1", interface=ImportInterface(
+                subroutine.symbol_table.lookup("b_mod"))))
+
+    # After parsing the a_1, a_2, b_1, b_2, not_used1 and not_used2 will have#
```
s/have#/have/

sergisiso replied:
Deleted #

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+    assert not isinstance(b_2, DataSymbol)
+
+    # Set up include_path to import the proper modules
+    Config.get()._include_paths = [str(tmpdir)]
```
Probably should use monkeypatch here to avoid polluting other tests.

sergisiso replied:
Oh yes, done!

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+        module b_mod
+            integer, parameter :: b_1 = 10
+            integer, save, pointer :: b_2
+            integer :: not_used1
```
Please could you add a private variable to this module and extend the test accordingly.

sergisiso replied:
Instead, I added a new test to better check the accessibility statements are respected, does it cover what you had in mind?

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+    assert isinstance(a_2, DataSymbol)
+    assert a_2.visibility == Symbol.Visibility.PRIVATE
+
+    # Clean up the config instance
```
monkeypatch will remove the need for this (and may also be better if the test should fail).

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+
+def test_resolve_imports_name_clashes(fortran_reader, tmpdir):
+    ''' Tests the SymbolTable resolve_imports method raises the appropriate
+    errors when it find name clashes. '''
```
"finds"!

sergisiso replied:
Fixed

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+    ''' Tests the SymbolTable resolve_imports method raises the appropriate
+    errors when it find name clashes. '''
+
+    with open(os.path.join(str(tmpdir), "a_mod.f90"), "w") as module:
```
pylint complains about these various `open` statements because they don't specify an encoding. Please could you add one.

sergisiso replied:
This is a new one :) Done

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+    symtab = subroutine.symbol_table
+
+    # Set up include_path to import the proper modules
+    Config.get()._include_paths = [str(tmpdir)]
```
monkeypatch please.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+    with pytest.raises(SymbolError) as err:
+        symtab.resolve_imports([symtab.lookup('a_mod')])
+    assert ("Found a name clash with symbol 'name_clash1' when importing "
+            "symbols from container 'a_mod', this symbol was already defined "
```
NIce.

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+
+
+def test_resolve_imports_with_datatypes(fortran_reader, tmpdir):
+    ''' Tests the SymbolTable resolve_imports method. '''
```
Please extend this docstring to distinguish this test from the first new one you added above.

sergisiso replied:
Done, also for other tests in this PR

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+        module my_mod
+            type my_type
+                integer :: field
+                integer, dimension(:,:) :: array
```
I think `array` must either be allocatable or have fixed dimensions.

sergisiso replied:
Added fixed dimensions, and since I was at it, I now assert that this dimension value is seen in the shape of the local symbol after resolving the import.

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+    assert not isinstance(symtab.lookup("other_type"), DataTypeSymbol)
+
+    # Set up include_path to import the proper modules and resolve symbols
+    Config.get()._include_paths = [str(tmpdir)]
```
monkeypatch please.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1643,3 +1646,232 @@ def test_rename_symbol():
+    Config.get()._include_paths = [str(tmpdir)]
+    symtab.resolve_imports()
+
+    # The global1 exist and is DataSymbols now
```
"...is a DataSymbol now"

sergisiso replied:
Done

arporter reviewed (changes requested):
New functionality looks OK and is well tested. However, I'm a bit worried that we might be biting off more than we can chew by automatically looking into all modules that we can find. Have you tried this functionality on e.g. the NEMO code? I'd be happier if we only did the resolving one module at a time until we found what we were looking for.
I think we may also need to deepcopy a symbol's datatype rather than the shallow copy that I think is happening at the moment.
I've checked all modifications for pylint - see inline.
Probably the Developer Guide would benefit from a description of this method (once we've agreed upon what it should do!).

sergisiso commented on src/psyclone/psyir/symbols/data_type_symbol.py:
```diff
@@ -107,6 +107,22 @@ def datatype(self, value):
+            raise TypeError("Argument should be of type 'DataTypeSymbol' but "
+                            "found '{0}'.".format(type(symbol_in).__name__))
+        super(DataTypeSymbol, self).copy_properties(symbol_in)
+        self._datatype = symbol_in.datatype
```
I am not totally sure on this one. If we made some modification inside a deep_copied type, then when generating the code it will not be reflected anywhere because the type is only described in the symbol definition (which in this case is in the external file) and they should match.

Also the INTEGER_TYPE, REAL_TYPE, ... that we have use a single instance of the specific scalar type and re-use the same instance everywhere.. which I think is more akind of a shallow reference for datatype.

Moreover, I see the types do not have a deepcopy method yet. So I would say it shouldn't be deepcopied for now, but probably at some point we need to design/think properly how types should behave as it has been done ad-hoc for now.

arporter replied:
OK, I think we need a discussion. My feeling was that a type should belong to a Symbol and not be shared but am happy to discuss. Look's like we'll be back to doing so on Teams though :-(

arporter replied:
(I'm happy to mark this as resolved for now.)

sergisiso commented on src/psyclone/psyir/symbols/data_type_symbol.py:
```diff
@@ -107,6 +107,22 @@ def datatype(self, value):
+
+        '''
+        if not isinstance(symbol_in, DataTypeSymbol):
+            raise TypeError("Argument should be of type 'DataTypeSymbol' but "
```
Done

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
 
+    def resolve_imports(self, container_symbols=None):
+        ''' Try to resolve deferred and unknown information from imported
+        symbols in this symbol table by searching for its definitions in
```
Done

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
+                # TODO #11: It would be useful to log this.
+                continue
+
+            for symbol in external_container.symbol_table.symbols:
```
Done

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
+                    continue  # We must ignore this symbol
+
+                if symbol.name in self:
+                    symbol_match = self.lookup(symbol.name)
```
Done (a slightly different comment)

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
+                            continue  # It doesn't come from this import
+
+                    # Found a match, update the interface if necessary or raise
+                    # and error if it is an ambiguous match
```
Fixed

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
+                    elif isinstance(interface, ImportInterface):
+                        if interface.container_symbol is not c_symbol:
+                            raise SymbolError(
+                                "Found a name clash with symbol '{0}' when "
```
Done

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
+                                "".format(symbol.name, c_symbol.name,
+                                          interface.container_symbol.name))
+                    else:
+                        raise SymbolError(
```
Done

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
+                            "'{1}'.".format(symbol.name, c_symbol.name))
+
+                    # Now copy the external symbol properties, but keep the
+                    # interface and visibility as this are local properties
```
Done

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
+                        else:
+                            symbol_match.specialise(type(symbol))
+                    symbol_match.copy_properties(symbol)
+                    symbol_match.interface = interface
```
Done

sergisiso commented on src/psyclone/tests/psyir/frontend/fparser2_test.py:
```diff
@@ -1578,6 +1578,41 @@ def test_use_stmt():
+
+
+def test_process_use_stmts_with_accessibility_statements(parser):
+    ''' Same than the previous test, but now from a module with a Fortran
```
Done

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
+                        # so it needs to be declared in the symbol table.
+                        new_symbol = symbol.copy()
+                        new_symbol.interface = ImportInterface(c_symbol)
+                        new_symbol.visibility = self.default_visibility
```
Yes the "private var" statement creates a generic var symbol with the appropriate visibility, so importing doesn't have to create a new symbol, just update the properties of the generic one. I added a new test to check this specific case.

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1067,6 +1069,93 @@ def copy_external_import(self, imported_var, tag=None):
                         " tag '{1}' is already used by another symbol."
                         "".format(imported_var, tag))
 
+    def resolve_imports(self, container_symbols=None):
```
I added an optional `target_symbol` argument to allow to call this method with a more constrained scope (but still allow to do a wildcard search if no target_symbol is specified as I think is the most similar to what the compiler would to).

I case of not finding the specified `target_symbol` I raise and error. This seemed appropriate if we are searching for a particular symbol, but let me know if you think differently.

sergisiso commented:
@arporter This is ready for the next review. I addressed all comments but the one on the datatype deepcopy which probably requires its own separate discussion.

sergisiso commented:
@arporter I took this out of review because when processing Nemo there is one issue remaining with the "wp" symbol getting the wrong visibility. This doesn't come from the resolve method because the visibility is already wrong in the external container. I think it is that the frontend has a separate handler to process precisions that may not be respecting the symbol table default visibility, but I will explore this as I would like this to work well for the whole Nemo code.

sergisiso commented:
@arporter Now I checked it works with Nemo (*at least the files not excluded by OpenMP/OpenACC) for a couple of dependencies I am interested, I had to make sure the symbol table default visibility are respected when parsing the external containers. Now I can do something like:
```
if "oce" in mod_sym_tab:
	oce_symbol = mod_sym_tab.lookup("oce")
	mod_sym_tab.resolve_imports(container_symbols=[oce_symbol])

if "dom_oce" in mod_sym_tab:
	oce_symbol = mod_sym_tab.lookup("dom_oce")
	mod_sym_tab.resolve_imports(container_symbols=[oce_symbol])
```
With this a 8-way parallel psyclone of all nemo source code takes: 56s
Without resolving wildcard imports I took: 48s

sergisiso commented:
@arporter This is ready for review again

arporter reviewed (changes requested):
I had a bit of trouble viewing the right diff of this PR this time around and also I got a bit delayed mid-review so I hope my comments still make sense.
There are just three comments left now.
I've taken the liberty of bringing this up-to-date with master.
Tests and examples are all OK with compilation.
All of the docs build OK.

sergisiso commented on src/psyclone/domain/lfric/psyir.py:
```diff
@@ -101,10 +101,10 @@
     # Create the specific symbol
     exec(
         "class {0}DataSymbol(DataSymbol):\n"
```
This was modified because these dynamically created LFRic symbol classes didn't have 'visibility' as a constructor argument. This argument should have been here anyway but this became an issue now because the new_symbol method assigns an explicit visibility to new symbols depending on the symbol table default visibility.

Since visibility (and the interface argument which has here before) are not responsibilities of this class, I opted for just a kwargs argument which is passed to the super constructor.

sergisiso replied:
I added a visibility argument on some of the tests for this classes

arporter replied:
Aha, thanks for explaining.

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1077,6 +1084,150 @@ def copy_external_import(self, imported_var, tag=None):
+            try:
+                external_container = c_symbol.container
+            # pylint: disable=broad-except
+            except Exception:
```
When I wrote this I wanted to capture any exception here, my take is:
The IOErrors happen often if the file is not found, but can also be ValueError is a file is found but inside doesn't contain the expected module. This are alright because we don't aim to find all dependencies (and could be named exceptions).

But the reason for a broad except is that then fparser can also (and does in Nemo) fail for some files and this can be NoMatchError, ParseError, FortranSyntaxError, ... We could chose to fail here if something of this happens if you prefer? A compiler would fail for any of these in a dependency. And we can target just specific modules to search for the wildcard import to circumvent these errors. But I wrote this more as a "do whatever it can" and if we find an error in some dependency just ignore this particular dependency and continue. I wouldn't list by name fparser errors because then it means updating/adding fparser errors means we need to update this (and we probably forget until it fails), plus the list of errors is becoming quite long.

I don't think this will contribute to the generic transformation scripts errors messages because this Exception is dropped here and not raised upwards, but of course logging it would be useful to know what had happen in each step of the dependency tree (well we still just do depth 1 for now).

arporter replied:
OK, and I see you have a TODO to add logging so that's fine.

sergisiso commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1077,6 +1084,150 @@ def copy_external_import(self, imported_var, tag=None):
+                        # Now we know where the symbol is coming from
+                        interface = ImportInterface(c_symbol)
+                    elif isinstance(interface, ImportInterface):
+                        if interface.container_symbol.name != c_symbol.name:
```
That's true. I updated the method and added a new tests for this specific case.

arporter replied:
Great

arporter commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -343,6 +345,11 @@ def new_symbol(self, root_name=None, tag=None, shadowing=False,
         else:
             symbol_type = Symbol
 
+        # If no visibility parameter has been provided use this symbol's table
```
"symbol table's"

arporter replied:
As this is just a comment I've fixed it myself.

arporter commented on src/psyclone/psyir/symbols/symboltable.py:
```diff
@@ -1077,6 +1084,149 @@ def copy_external_import(self, imported_var, tag=None):
+        referred external container. A single symbol to resolve can be
+        specified for a more targeted import.
+
+        :param container_symbols: list of container symbols to search in
```
Missing line continuation.

arporter replied:
As this is just a docstring I've fixed it myself.

arporter commented on src/psyclone/tests/psyir/symbols/symboltable_test.py:
```diff
@@ -1504,6 +1507,24 @@ def test_new_symbol():
 
 
+def test_new_symbol_with_private_default_visibility():
+    '''Test that the new_symbol method creates method with the appropriate
```
"...creates a symbol with the appropriate..."

arporter replied:
Just a docstring so fixed.

arporter reviewed (approved):
I've taken the liberty of fixing a few typos in comments/docstrings
All requested changes have been made.
All tests and examples are OK with compilation.
All docs build OK.
